### DISCUSSION TITLE
Backwards compability of .cjs and .mjs files - How are they handled in older Node.js versions?

### DISCUSSION THREAD

**Discussion by advename** [nodejs/help#4898] [Q&A]
Following the [documentation](https://nodejs.org/api/packages.html#determining-module-system), lets say that since Node.js v12 we can be more specific about ES or CJS modules with:
- using the `.cjs` file extension
- using the `.mjs` file extension
- setting the `type` field in `package.json`

However, one thing that i failed to find and am curious about - how are these file extensions handled in Node.js versions prior to 12?

**Comment by Trott** [comment]
Prior to Node.js 12, all files were treated as CommonJS files by default. ESM was not supported (outside of the `--experimental-modules` flag).

**Reply by advename** [reply]
That I'm aware of. What I meant is, what happens when I run a `.cjs` file in e.g. Node.js v.8? Does it throw an error since it only understands `.js` files, or is there anything going on making it run on a level i have no knowledge of?

**Reply by Trott** [reply]
As I wrote:

> Prior to Node.js 12, all files were treated as CommonJS files by default.

It depends on the contents of the file, not so much the extension. If you happen to write CommonJS code and put it in a `.mjs` file, it will work fine in Node.js prior to 12 (but not in Node.js 12 and after).

Using Node.js 10.x, a file that consists of `import "fs"` throws a `SyntaxError` (unless `--experimental-modules` flag is used). This is true whether the file extension is `.js`, `.mjs`, or `.cjs`. It doesn't matter. If the same file contains `require('fs')`, it will work just fine in Node.js 10.x no matter which of those three extensions it uses. (It will throw a `SyntaxError` with a `.json` extension though.)

> Does it throw an error since it only understands .js files

As far as I know, Node.js prior to 12.x doesn't pay any attention at all to file extensions (excluding use of the `--experimental-modules` flag) other than `.json` and maybe `.node`.

If you use a Node.js version manager like [`nvm`](https://github.com/nvm-sh/nvm), you can switch versions easily and test out whatever it is you want to know for yourself.
